core/services/run_queue.go

```diff
@@ -56,24 +57,27 @@ func (rq *runQueue) Start() error {
 
 // Stop closes all open worker channels.
 func (rq *runQueue) Stop() {
+	rq.workersMutex.Lock()
+	rq.stopRequested = true
+	rq.workersMutex.Unlock()
 	rq.workersWg.Wait()
 }
 
 // Run tells the job runner to start executing a job
 func (rq *runQueue) Run(run *models.JobRun) {
-	runID := run.ID.String()
+	rq.workersMutex.Lock()
+	if rq.stopRequested {
+		return
+	}
```

Comment:
It looks like the above enters into a weird deadlock even if 'stop has been requested'.
Why no use defer?
```
 	rq.workersMutex.Lock()
        defer rq.workersMutex.Unlock()
```

Reply:
Good catch @felder-cl

Reply:
IIRC, The Unlock needs to happen before the `Add` otherwise it can cause a deadlock.